File: ledger-models-java/src/test/java/bench/LazyHydrateBench.java
package bench;

import common.models.transaction.Transaction;
import common.util.LinkCache;
import fintekkers.models.portfolio.PortfolioProto;
import fintekkers.models.transaction.TransactionProto;
import fintekkers.models.util.LocalTimestamp.LocalTimestampProto;
import fintekkers.models.util.Uuid.UUIDProto;
import org.junit.jupiter.api.Test;
import protos.serializers.util.proto.ProtoSerializationUtil;

import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * End-to-end perf bench for Transaction wrapper lazy hydration. Build N
 * link-mode {@link TransactionProto}s, pre-warm {@link LinkCache#TRANSACTION}
 * + {@link LinkCache#PORTFOLIO}, then construct N wrappers and read accessors.
 *
 * <p>Run: {@code ./gradlew test --tests bench.LazyHydrateBench}. Output goes
 * to stdout; not an assertion-based test.
 */
class LazyHydrateBench {

    private static final int[] SIZES = {10, 100, 1_000, 10_000};

    @Test
    void bench_transaction_lazy_hydrate() {
        System.out.println("# java bench: lazy-hydrate Transaction via pre-warmed LinkCache");
        for (int n : SIZES) {
            runBench(n);
        }
    }

    private static void runBench(int n) {
        ZonedDateTime asOf = ZonedDateTime.now().withNano(0);
        LocalTimestampProto asOfProto = ProtoSerializationUtil.serializeTimestamp(asOf);

        List<UUID> txnUuids = new ArrayList<>(n);
        List<UUID> portUuids = new ArrayList<>(n);
        List<TransactionProto> linkProtos = new ArrayList<>(n);

        for (int i = 0; i < n; i++) {
            UUID txnUuid = UUID.randomUUID();
            UUID portUuid = UUID.randomUUID();
            txnUuids.add(txnUuid);
            portUuids.add(portUuid);

            UUIDProto txnUuidProto = ProtoSerializationUtil.serializeUUID(txnUuid);
            UUIDProto portUuidProto = ProtoSerializationUtil.serializeUUID(portUuid);

            PortfolioProto resolvedPortfolio = PortfolioProto.newBuilder()
                    .setUuid(portUuidProto)
                    .setAsOf(asOfProto)
                    .setIsLink(false)
                    .setPortfolioName("P-" + portUuid.toString().substring(0, 8))
                    .build();
            TransactionProto resolved = TransactionProto.newBuilder()
                    .setUuid(txnUuidProto)
                    .setAsOf(asOfProto)
                    .setIsLink(false)
                    .setTradeName("T-" + txnUuid.toString().substring(0, 8))
                    .setPortfolio(resolvedPortfolio)
                    .build();
            LinkCache.TRANSACTION.put(txnUuid, resolved, asOf);
            LinkCache.PORTFOLIO.put(portUuid, resolvedPortfolio, asOf);

            linkProtos.add(TransactionProto.newBuilder()
                    .setUuid(txnUuidProto)
                    .setAsOf(asOfProto)
                    .setIsLink(true)
                    .build());
        }

        System.gc();
        long heapBefore = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        long t0 = System.nanoTime();
        int sink = 0;
        for (TransactionProto link : linkProtos) {
            Transaction txn = new Transaction(link);
            String name = txn.getTradeName();
            if (name != null && !name.isEmpty()) sink++;
        }
        long t1 = System.nanoTime();
        long heapAfter = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

        double elapsedMs = (t1 - t0) / 1_000_000.0;
        double perOpUs = ((double) (t1 - t0) / n) / 1_000.0;
        double heapDeltaKb = (heapAfter - heapBefore) / 1024.0;

        System.out.printf(
                "N=%6d  elapsed=%9.2f ms  per_op=%8.2f us  heap_delta=%9.2f KiB  reads=%d%n",
                n, elapsedMs, perOpUs, heapDeltaKb, sink);

        for (UUID u : txnUuids) LinkCache.TRANSACTION.evict(u);
        for (UUID u : portUuids) LinkCache.PORTFOLIO.evict(u);
    }
}

File: ledger-models-rust/tests/lazy_hydrate_bench.rs
//! End-to-end perf bench for TransactionWrapper lazy hydration.
//!
//! Run: `cargo test --release --test lazy_hydrate_bench bench_transaction_lazy_hydrate -- --nocapture`
//! (release mode + --nocapture to see stdout).
//!
//! Same shape as the Java/Python/TS benches — pre-warm LinkCache.TRANSACTION
//! and LinkCache.PORTFOLIO with N resolved protos, then construct N
//! TransactionWrapper instances from link-mode protos and read accessors.

use ledger_models::fintekkers::models::portfolio::PortfolioProto;
use ledger_models::fintekkers::models::transaction::TransactionProto;
use ledger_models::fintekkers::models::util::{LocalTimestampProto, UuidProto};
use ledger_models::fintekkers::wrappers::models::transaction::TransactionWrapper;
use ledger_models::fintekkers::wrappers::util::link_cache;
use prost_types::Timestamp;
use std::time::Instant;
use uuid::Uuid;

const SIZES: &[usize] = &[10, 100, 1_000, 10_000];

fn make_as_of() -> LocalTimestampProto {
    LocalTimestampProto {
        timestamp: Some(Timestamp {
            seconds: 1_700_000_000,
            nanos: 0,
        }),
        time_zone: "UTC".to_string(),
    }
}

fn run_bench(n: usize) {
    let as_of = make_as_of();

    let mut txn_uuids: Vec<Uuid> = Vec::with_capacity(n);
    let mut port_uuids: Vec<Uuid> = Vec::with_capacity(n);
    let mut links: Vec<TransactionProto> = Vec::with_capacity(n);

    for _ in 0..n {
        let txn_uuid = Uuid::new_v4();
        let port_uuid = Uuid::new_v4();
        txn_uuids.push(txn_uuid);
        port_uuids.push(port_uuid);

        let resolved_portfolio = PortfolioProto {
            uuid: Some(UuidProto {
                raw_uuid: port_uuid.as_bytes().to_vec(),
            }),
            as_of: Some(as_of.clone()),
            is_link: false,
            portfolio_name: format!("P-{}", &port_uuid.to_string()[0..8]),
            ..Default::default()
        };
        let resolved_txn = TransactionProto {
            uuid: Some(UuidProto {
                raw_uuid: txn_uuid.as_bytes().to_vec(),
            }),
            as_of: Some(as_of.clone()),
            is_link: false,
            trade_name: format!("T-{}", &txn_uuid.to_string()[0..8]),
            portfolio: Some(resolved_portfolio.clone()),
            ..Default::default()
        };
        link_cache::transaction().put(txn_uuid, resolved_txn, Some(as_of.clone()));
        link_cache::portfolio().put(port_uuid, resolved_portfolio, Some(as_of.clone()));

        links.push(TransactionProto {
            uuid: Some(UuidProto {
                raw_uuid: txn_uuid.as_bytes().to_vec(),
            }),
            as_of: Some(as_of.clone()),
            is_link: true,
            ..Default::default()
        });
    }

    let t0 = Instant::now();
    let mut sink = 0usize;
    for link in &links {
        let w = TransactionWrapper::new(link.clone());
        if !w.trade_name().is_empty() {
            sink += 1;
        }
        if w.portfolio().is_some() {
            sink += 1;
        }
    }
    let elapsed = t0.elapsed();
    let per_op_us = elapsed.as_nanos() as f64 / n as f64 / 1_000.0;

    println!(
        "N={:>6}  elapsed={:>9.2} ms  per_op={:>8.2} us  reads={}",
        n,
        elapsed.as_secs_f64() * 1000.0,
        per_op_us,
        sink
    );

    for u in &txn_uuids {
        link_cache::transaction().evict(*u);
    }
    for u in &port_uuids {
        link_cache::portfolio().evict(*u);
    }
}

#[test]
fn bench_transaction_lazy_hydrate() {
    println!("# rust bench: lazy-hydrate Transaction via pre-warmed LinkCache");
    for n in SIZES {
        run_bench(*n);
    }
}
